module/validation/receipt_validator.go

```diff
@@ -160,64 +161,203 @@ func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
 	// reference each other.
 	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
 	// Eventually invalid result will be detected and fail the whole validation.
-	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
-		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[previousResultID]
 		if found {
 			return prevResult, nil
 		}
 
-		return v.previousResult(executionResult)
+		return v.fetchResult(previousResultID)
 	}
 
 	for i, r := range receipts {
-		err := v.validate(r, previousResult)
+		prevResult, err := fetchResult(r.ExecutionResult.PreviousResultID)
+		if err != nil {
+			return err
+		}
+
+		err = v.validateReceipt(r.Meta(), r.ExecutionResult.BlockID)
```

Comment:
Do we want to check if we have this result in the storage already, in which case means we have validated already, and can skip?

so that I will only validate results if I have never seen them before

Reply:
This function accepts `ExecutionReceipt` as argument so it  means that it's a receipt meta + result, this can happen when receipt was introduced as single entity. So it can't be that it's referring another result which was introduced in previous block. It should be result that was never seen before.

Reply:
I think it still could. for instance, a consensus node receives a receipt for block A from EN1 and put it in block B, and then another consensus node receives block B and also receives another receipt for blockA from EN2, so it will validate this receipt which has been added to any block and whose result has been added to block B

Reply:
wait, but this function is called only when we got a receipt from execution node, not when it was included in a block

Reply:
Yes. So when the matching engine receipts a full receipt from execution node, it will validate the receipt with the `Validate` function, right? And you mentioned:

>  It should be result that was never seen before.

Which I don't think so, because it's possible we receive a receipt whose result is known, and has been added to previous block, and we could skip the validation on the result since we must have validated that result before accepting the block.